File: haskell-server/src/Bolour/Plane/Domain/Axis.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}

{-|
X and Y axis and related types.
-}
module Bolour.Plane.Domain.Axis (
    Axis(..)
  , crossAxis
  , Coordinate
  , Height
  , Width
  , Direction
  , forward
  , backward
) where

import GHC.Generics (Generic)
import Control.DeepSeq (NFData)
import Data.Aeson (FromJSON, ToJSON)

-- | One of the axes of the board.
--   X designates a horizontal line of the board.
--   An index value associated with X is a column number. A size is the width.
--   Y designates a vertical line of the board.
--   An index value associated with Y is a row number. A size the height.
--
--   Note however that the first coordinate of a Point is a row, and its second
--   coordinate is a column. That is probably a mistake since it is inconsistent
--   with the use of axes. But for historical reasons still remains in the code base.
data Axis = Y | X
  deriving (Eq, Show, Generic)

instance FromJSON Axis
instance ToJSON Axis

crossAxis :: Axis -> Axis
crossAxis X = Y
crossAxis Y = X

type Direction = Int

-- | Increment for going forward in a line.
forward :: Int
forward = 1
-- | Increment for going backward in a line.
backward :: Int
backward = -1

-- | A value of a board coordinate.
type Coordinate = Int
type Height = Coordinate
type Width = Coordinate

File: scala-server/plane/src/main/scala/com/bolour/plane/domain/Axis.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */
package com.bolour.plane.domain

/**
  * The X or Y axis of the plane.
  *
  * TODO. Enumeration is 'deprecated'. Use sealed abstract case class/objects.
  */
object Axis extends Enumeration {
  type Axis = Value
  val X, Y = Value

  def crossAxis(axis: Axis): Axis = if (axis == X) Y else X

  val forward = +1
  val backward = -1

}

